### PR TITLE
[WIP] Conversion Script for DiffusionDet

### PR DESCRIPTION
**Pull request by 345ishaan**
Create a script which builds a SwinModel from HuggingFace transformers and load weights from pretrained checkpoints provided by DiffusionDet author's.

More details in: https://github.com/huggingface/diffusers/issues/1350

NOTE: This is WIP as I still need to build the pipeline and handle decoder.

### CODE REVIEW

File: scripts/convert_diffusion_det_to_diffusers.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      _load_weights(f'backbone.bottom_up.layers.{layer_id}.downsample.reduction.weight', f'encoder.layers.{layer_id}.downsample.reduction.weight')
+      _load_weights(f'backbone.bottom_up.layers.{layer_id}.downsample.norm.weight', f'encoder.layers.{layer_id}.downsample.norm.weight')
+      _load_weights(f'backbone.bottom_up.layers.{layer_id}.downsample.norm.bias', f'encoder.layers.{layer_id}.downsample.norm.bias')
+
+  _load_swin_layers(layer_id=0, num_blocks=2)
+  _load_swin_layers(layer_id=1, num_blocks=2)
+  _load_swin_layers(layer_id=2, num_blocks=18)
+  _load_swin_layers(layer_id=3, num_blocks=2, downsample=False)
+
+  # TODO(ishan): check absent layernorm's after layer-{0,1,2}
```
Comment (345ishaan) [inline scripts/convert_diffusion_det_to_diffusers.py:99]: @patrickvonplaten @ShoufaChen  can you help shed some light on this?

I am seeing the provided checkpoint has layernorm after every layer as i see in the weights:
``
backbone.bottom_up.norm0.weight torch.Size([128])
backbone.bottom_up.norm0.bias torch.Size([128])
backbone.bottom_up.norm1.weight torch.Size([256])
backbone.bottom_up.norm1.bias torch.Size([256])
backbone.bottom_up.norm2.weight torch.Size([512])
backbone.bottom_up.norm2.bias torch.Size([512])
backbone.bottom_up.norm3.weight torch.Size([1024])
backbone.bottom_up.norm3.bias torch.Size([1024])
```
however, the corresponding model which i loading from HF config corresponding to `microsoft/swin-base-patch4-window7-224-in22k` only has layernorm after the last layer.

**Comment by patrickvonplaten** [comment]
Very cool progress @345ishaan :-)

**Comment by 345ishaan** [comment]
I looked at the scheduler part in the original code. The scheduling logic looks similar to the DDIM scheduler in pipleines.
will look into the decoder part to see if it can be utilised by existing HF implementation.

@patrickvonplaten regarding the NMS postprocessing, what do you suggest? anything i can utilise else i was planning to create a util by forking from detectron code unless you are okay adding detectron2 dependency i.e `from detectron2.layers import batched_nms`

**Comment by patrickvonplaten** [comment]
@345ishaan,

Definitely happy to add detectron as an optional dependency as e.g. done here: https://github.com/huggingface/diffusers/blob/135567f18edc0bf02d515d5c76cc736d1ebddad3/src/diffusers/pipelines/__init__.py#L29

e.g. we would have something like: 

```python
try:
    if not (is_torch_available() and is_detectron_available()):
        raise OptionalDependencyNotAvailable()
except OptionalDependencyNotAvailable:
    from ..utils.dummy_torch_and_detectron_objects import *  # noqa F403
else:
    from .diffusion_det import DiffusionDetPipeline
```

**Comment by patrickvonplaten** [comment]
Let me know if this doesn't make sense :-)

**Comment by 345ishaan** [comment]
Sorry for the delay here. I still plan to work it.

**Pull request closed without merging**